File: algorithms/merge-sort/merge-sort.c
/*
* 
* Bojan Skrchevski 2018-03-28
*
* Merge sort algorithm implementation in C
*
*/

#include<stdio.h>


/*
* 
* Merge method, implemented with a temporarty array
* 
*/
void merge(int array[], int start, int mid, int end)
{
    int i = 0, j, ll = start;
    int second_half_start = mid + 1;
    int tmpArray[100];
    
    while((start <= mid) && (second_half_start <= end)) {
        // in case both arrays have values
        // compare the first two elements
        // assign the left or the right first 
        // item to the temp array 
        // (assign and increment the counters in one step)
        if(array[start] < array[second_half_start]) {
            tmpArray[i++] = array[start++];
        } else {
            tmpArray[i++] = array[second_half_start++];
        }
    }
    
    // in case we only have the left part of the array
    while(start <= mid) {
        tmpArray[i++] = array[start++];
    }
    
    // in case we only have the right part of the array
    while(second_half_start <= end) {
        tmpArray[i++] = array[second_half_start++];
    }
    
    // reassign the sorted values from the 
    // temporary array to the original array
    for(j=0; j < i; j++) {
        array[ll++] = tmpArray[j];
    }
}

/*
*
* Merge sort method.
*
* Recursively splits the array in halves and calls the merge
* method once it reaches size the end.
* 
*/
void merge_sort(int array[], int start, int end)
{
    int mid;

    if(start < end)
    {
        mid = (end + start) / 2;
        merge_sort(array, start, mid);
        merge_sort(array, mid + 1, end);
        merge(array, start, mid, end);
    }
}

void main()
{
    int array[100];
    int n, i;
    
    printf("\nNum of elements present in the array? :");
    scanf("%d", &n);
    
    for(i=0; i < n; i++)
    {
        scanf("%d", &array[i]);
    }
    
    merge_sort(array, 0, n-1);
    
    printf("\nThe sorted array is...\n");
    for(i=0; i < n; i++)
    {
        printf("%d ", array[i]);
    }

    printf("\n\n");
}

File: algorithms/merge-sort/merge-sort.cs
/*
* 
* Bojan Skrchevski 2018-03-18
*
* Merge sort algorithm.
*
*
*/

using System;
using System.Linq;
using System.Collections.Generic;

class MergeSortAlgorithm
{
    static int[] MergeSort(int[] array) 
    {   
        if(array.Length <= 1) {
            return array;
        }

        int middle = array.Length / 2;
        int[] left = MergeSort(array, 0, middle);
        int[] right = MergeSort(array, middle + 1, array.Length - 1);
        
        return Merge(left, right);
    }

    static int[] MergeSort(int[] array, int start, int end) 
    {
        if(array.Length <= 1) {
            return array;
        }

        int middle = array.Length / 2;
        int[] left = MergeSort(array, start, middle);
        int[] right = MergeSort(array, middle + 1, end);
        
        return Merge(left, right);
    }

    static int[] Merge(int[] left, int[] right)
    {
        List<int> result = new List<int>();
        List<int> lefty = new List<int>(left);
        List<int> righty = new List<int>(right);

        while(lefty.Count > 0 || righty.Count > 0) {
            if(lefty.Count > 0 && righty.Count > 0) {
                //if the first item on left is
                //less than right...
                if(lefty[0] < righty[0]) {
                    //take the first item on the left
                    result.Add(lefty.First());
                    lefty.Remove(lefty.First());
                } else {
                    //take the first item on the right
                    result.Add(righty.First());
                    righty.Remove(righty.First());
                }
            } else if(lefty.Count > 0) {
                result.Add(lefty.First());
                lefty.Remove(lefty.First());
            } else {
                result.Add(righty.First());
                righty.Remove(righty.First());
            }
        }

        return result.ToArray();
    }

    static void Main(String[] args) {
        
        string[] a_temp = Console.ReadLine().Split(' ');
        int[] array = Array.ConvertAll(a_temp, Int32.Parse);

        int[] sorted = MergeSort(array);

        foreach (var item in sorted)
        {
            Console.Out.Write(item + " ");
        }

        Console.Out.WriteLine();
        Console.ReadKey();
    }
}